[REBASE & FF] NvmExpressDxe: Request Number of Queues from Controller

VivianNK opened the pull request:
## Description

Request Number of Queues from the Controller. If the number of queues is <2, then we only have a synchronous queue and do not support asynchronous BlockIo2 functionality.

Created issue https://github.com/microsoft/mu_basecore/issues/1335 for `PcdSupportAlternativeQueueSize `

Created issue https://github.com/microsoft/mu_basecore/issues/1358 for continuing after failing to install BlockIo2

- [x] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?
- [ ] Backport to release branch?

## How This Was Tested

CI and QEMU with NVMe virtualization with max_ioqpairs=1, 2, 64

Used `reconnect` and `connect -r` cmds in Shell on the virtualized NVMe controller handle to trigger the `NvmeControllerReset` function (saw in the run log).
Confirmed the device was reconnected using `devtree`.
Successfully wrote to file systems on the NVMe drive to test data queues. 

## Integration Instructions

N/A

chris-oo committed:
MdeModulePkg: Add alternative queue sizes in nvme driver to support specific devices

Add a new PCD that toggles between the old driver behavior and new behavior that uses a different hardcoded queue size to support specific devices.  Default behavior should not change.

MdeModulePkg/NvmExpressDxe: Improve NVMe controller init robustness

It has been observed that some faulty NVMe devices may return
invalid values. The code in NvmExpressDxe recognizes the controller
is not responding correctly and issues an ASSERT() often in DEBUG
builds or a reset in RELEASE builds.

The following changes are made to NvmeControllerInit() to prevent a
faulty NVMe device from halting the overall boot:

1. Check the Vendor ID and Device ID to verify they are read properly
   and if not, return EFI_DEVICE_ERROR
2. Replace the ASSERT() when Memory Page Size Minimum (Cap.Mpsmin)
   with an error message and return EFI_DEVICE_ERROR

Signed-off-by: Michael Kubacki <michael.kubacki@microsoft.com>

MdeModulePkg/NvmExpressDxe: Correct function parameter modifer

Updates the `Cap` parameter for `ReadNvmeControllerCapabilities()`
to be `OUT` since the buffer pointed to is written within the
function.

Signed-off-by: Michael Kubacki <michael.kubacki@microsoft.com>

VivianNK committed:
MdePkg: Nvme.h add unions for Number of Queues

Added unions to support using the Set Feature
Command for Number of Queues.

Signed-off-by: Vivian Nowka-Keane <vnowkakeane@linux.microsoft.com>

VivianNK committed:
MdeModulePkg: NvmExpressDxe: Request queues from
the controller.

Instead of assuming we always want to create two sets of data queues,
we request two and the controller will determine how many are allocated.

Signed-off-by: Vivian Nowka-Keane <vnowkakeane@linux.microsoft.com>

VivianNK commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -172,15 +184,17 @@ EnumerateNvmeDevNamespace (
-    Device->BlockIo2.WriteBlocksEx = NvmeBlockIoWriteBlocksEx;
-    Device->BlockIo2.FlushBlocksEx = NvmeBlockIoFlushBlocksEx;
-    InitializeListHead (&Device->AsyncQueue);
+    if (Private->NSQA > 1 || Private->NCQA > 1) {
```
Is it necessary to check NCQA > 1?

Will NCQA > NSQA ever be true?

VivianNK replied:
Assuming NSQA=1 and NCQA=2 is possible, would we want to install BlockIo2? @apop5

VivianNK replied:
Further, what about NSQA=2 and NCQA=1? That's not a case I've seen but one I know is possible.

VivianNK replied:
I can change the check to && instead of || so there are at least 2 of each before installing BlockIo2, I'm just unsure of the expected behavior of the above cases. 
| # of CQ   | # of SQ    | Behavior |
| ------------- | ------------- | ----------------|
1 | 1 | Blocking Data queue only. Don't install BlockIo2
1 | 2+ | I can see a scenario where this uses both blocking and unblocking - ?
2 +| 1 | I Assume blocking only?
2+ | 2+ | Blocking and unblocking data queue. Install BlockIo2

apop5 replied:
even though it was mentioned that submission/completion queues do not have to match, for our purposes in firmware, we should treat them as needing to match.

I don't think it is worth codifying mis matched queues, I think it is safe to have the following:
Admin Submission/Completion - Always required
Submission/Completion queues 1 - Required if we want to Support BlockIo. If there is not at least 1 submission/completion queue available, then we dont have a way to communicate to the drive. 
Submission/Completion queues 2 - Install BlockIo and BlockIo2

If you want to collapse NSQA and NCQA into one variable, that can work  as well. It would tell you if BlockIo2 is supported.

VivianNK commented:
[like]  Vivian Nowka-Keane reacted to your message:
________________________________
From: Aaron ***@***.***>
Sent: Monday, February 3, 2025 10:43:11 PM
To: microsoft/mu_basecore ***@***.***>
Cc: Vivian Nowka-Keane ***@***.***>; Author ***@***.***>
Subject: Re: [microsoft/mu_basecore] NvmExpressDxe: Request Number of Queues from Controller (PR #1260)


@apop5 commented on this pull request.

________________________________

In MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c<https://github.com/microsoft/mu_basecore/pull/1260#discussion_r1940228636>:

> @@ -172,15 +184,17 @@ EnumerateNvmeDevNamespace (
     Device->BlockIo.WriteBlocks = NvmeBlockIoWriteBlocks;
     Device->BlockIo.FlushBlocks = NvmeBlockIoFlushBlocks;

-    //
-    // Create BlockIo2 Protocol instance
-    //
-    Device->BlockIo2.Media         = &Device->Media;
-    Device->BlockIo2.Reset         = NvmeBlockIoResetEx;
-    Device->BlockIo2.ReadBlocksEx  = NvmeBlockIoReadBlocksEx;
-    Device->BlockIo2.WriteBlocksEx = NvmeBlockIoWriteBlocksEx;
-    Device->BlockIo2.FlushBlocksEx = NvmeBlockIoFlushBlocksEx;
-    InitializeListHead (&Device->AsyncQueue);
+    if (Private->NSQA > 1 || Private->NCQA > 1) {


even though it was mentioned that submission/completion queues do not have to match, for our purposes in firmware, we should treat them as needing to match.

I don't think it is worth codifying mis matched queues, I think it is safe to have the following:
Admin Submission/Completion - Always required
Submission/Completion queues 1 - Required if we want to Support BlockIo. If there is not at least 1 submission/completion queue available, then we dont have a way to communicate to the drive.
Submission/Completion queues 2 - Install BlockIo and BlockIo2

If you want to collapse NSQA and NCQA into one variable, that can work as well. It would tell you if BlockIo2 is supported.

—
Reply to this email directly, view it on GitHub<https://github.com/microsoft/mu_basecore/pull/1260#discussion_r1940228636>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AEFBXXPOG5SGJ27MEM5AP7D2N7WH7AVCNFSM6AAAAABVYVQGCOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKOJRGMYDKMJTGA>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

Flickdm commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -172,14 +172,17 @@ EnumerateNvmeDevNamespace (
-    Device->BlockIo2.ReadBlocksEx  = NvmeBlockIoReadBlocksEx;
-    Device->BlockIo2.WriteBlocksEx = NvmeBlockIoWriteBlocksEx;
-    Device->BlockIo2.FlushBlocksEx = NvmeBlockIoFlushBlocksEx;
+    if ((Private->NSQA > 1) && (Private->NCQA > 1)) {
```
What is 1 and can we make it a define?

VivianNK replied:
This is an outdated version of the file and a define is used here: https://github.com/microsoft/mu_basecore/pull/1260/files#diff-2621270f638ebef3e2c838a721daa614432066ba848b199d4cfb096ef27b6d41R176

Flickdm commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -271,6 +272,19 @@ EnumerateNvmeDevNamespace (
+      Status = gBS->InstallProtocolInterface (
+                      &Device->DeviceHandle,
+                      &gEfiBlockIo2ProtocolGuid,
+                      EFI_NATIVE_INTERFACE, // this is the only option in the enum..?
```
Is this a question about what this should be?

VivianNK replied:
Outdated file and has since been removed

spbrogan commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -271,6 +271,19 @@ EnumerateNvmeDevNamespace (
+      Status = gBS->InstallProtocolInterface (
+                      &Device->DeviceHandle,
+                      &gEfiBlockIo2ProtocolGuid,
+                      EFI_NATIVE_INTERFACE, // this is the only option in the enum..?
```
this comment should be removed before completed

spbrogan commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -172,14 +172,16 @@ EnumerateNvmeDevNamespace (
-    Device->BlockIo2.ReadBlocksEx  = NvmeBlockIoReadBlocksEx;
-    Device->BlockIo2.WriteBlocksEx = NvmeBlockIoWriteBlocksEx;
-    Device->BlockIo2.FlushBlocksEx = NvmeBlockIoFlushBlocksEx;
+    if (Private->NSQA > 1) {
```
unfortunately we should surround all these changes with a MU_CHANGE comment tag until this goes upstream

spbrogan commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -271,6 +271,19 @@ EnumerateNvmeDevNamespace (
+                      &Device->BlockIo2
+                      );
+      if (EFI_ERROR (Status)) {
+        goto Exit;
```
I know the existing code doesn't do this but i would suggest a debug statement to indicate the error path

spbrogan commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -506,6 +525,28 @@ UnregisterNvmeNamespace (
     return Status;
   }
 
+  //
```
this logic flow might not be correct.  

What is the point of the open protocol on error?  Does this signal the unregister failed?

VivianNK replied:
I think it signals the failure somehow, it's the same flow followed by an error on uninstallMultipleProtocolInterfaces on line # 517 and on uninstall of the Storage Security Command Protocol on line # 567. I looked at git blame and that behavior has been there since this file was added:
https://github.com/microsoft/mu_basecore/commit/eb290d0257814701eefd93ee71d46efb6dcc320b#diff-2621270f638ebef3e2c838a721daa614432066ba848b199d4cfb096ef27b6d41R370

Who might know more about this?

apop5 replied:
Are you saying that the CloseProtocol on line 494 needs to be moved to after successfully uninstalling all protocols on the child handle?

apop5 replied:
if a single uninstall protocol fails in UninstallMultipleProtocolInterfaces, all the protocols will be reinstalled on the handle and EFI_INVALID_PARAMTER will be returned. (https://github.com/microsoft/mu_basecore/blob/dev/202405/MdeModulePkg/Core/Dxe/Hand/Handle.c#L841)

In the case that the uninstall fails, the hierarchy between the controller and the child handle needs to be reestablished to allow the system to see the connection between the two.

spbrogan commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -1112,24 +1153,27 @@ NvmExpressDriverBindingStart (
     //
     // Start the asynchronous I/O completion monitor
     //
```
I would suggest a nice comment here to indicate that the event and timer are only used for blockio2 and that is only used when the number of queues is greater than 1.

spbrogan commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.h:
```diff
@@ -78,7 +78,8 @@ extern EFI_DRIVER_SUPPORTED_EFI_VERSION_PROTOCOL  gNvmExpressDriverSupportedEfiV
 
-#define NVME_MAX_QUEUES  3                              // Number of queues supported by the driver
+// Maximum number of queue pairs supported by the driver, including the admin queues.
+#define NVME_MAX_QUEUES  3
```
additional comment here would be good about scenario where only 2 queues are supported and the field that should be used instead of this to determine how many queues are active.

spbrogan reviewed (commented):
Didn't finish review but here are few comments

apop5 commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.h:
```diff
@@ -161,6 +177,18 @@ struct _NVME_CONTROLLER_PRIVATE_DATA {
+  //
+  // Number of Queues Allocated by the controller
+  //
+  UINT32                                Nsqa; // Number of Submission Queues Allocated
```
The above checks are only checking Nsqa.

Uefi always uses a 1:1 submission/completion queue allocation. Is there any benefit to having both Ncqa and Nsqa? Would it be better to just track a single SubmissionCompletionPairCount?

VivianNK replied:
Because there can be multiple submission queues for 1 completion queue, and the possibility of that scenario doesn't seem far off, it made sense to me to check only Nsqa. Since in that case we would have 1 completion queue, 1 synchronous submission queue, and 1 asynchronous submission queue. 
The goal here was to be more flexible for future scenarios. It may be unnecessary for this PR, though. Will start a private thread for discussion.

VivianNK replied:
@apop5  Changed to `NumberOfDataQueuePairs` since the consensus was, we only support 1:1 submission: completion queues.

apop5 commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -1194,7 +1297,7 @@ NvmExpressDriverBindingStart (
   if ((Private != NULL) && (Private->Buffer != NULL)) {
     // MU_CHANGE - Support alternative hardware queue sizes in NVME driver
-    PciIo->FreeBuffer (PciIo, QueuePageCount, Private->Buffer);
+    PciIo->FreeBuffer (PciIo, AdminQueuePageCount, Private->Buffer); // MU_CHANGE - Allocate IO Queue Buffer
```
Unchecked status here that will eventually be picked up by codeql.

```suggestion
    Status = PciIo->FreeBuffer (PciIo, AdminQueuePageCount, Private->Buffer); // MU_CHANGE - Allocate IO Queue Buffer
    if( EFI_ERROR(Status)) {
      DEBUG(DEBUG_ERROR, "%a() - Failed to Free AdminQueuePageCount %r\n", __func__, Status));
    }

```

apop5 commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -718,29 +890,171 @@ NvmeCreateIoSubmissionQueue (
+  UINTN       Index;
+  EFI_STATUS  Status;
+
+  // TODO do we want to store the AsqsPages/QueuePairPageCount in the private data structure? so we don't have to recaclulate it later on?
```
I think you can drop the todos.

While it would speed up execution, the speed up would be minimal

apop5 commented on MdePkg/Include/IndustryStandard/Nvme.h:
```diff
@@ -984,6 +1005,30 @@ typedef enum {
   SanitizeOpcode                = NVME_ADMIN_SANITIZE_CMD
 } NVME_ADMIN_COMMAND_OPCODE;
 
+// MU_CHANGE [BEGIN] - Add Nvm Express Admin Feature Identifiers
```
These defines are coming from Nvme specification.

They should probably be moved into Mde\Include\IndustryStandard\Nvme.h for public reuse.

apop5 commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -1318,11 +1421,26 @@ NvmExpressDriverBindingStop (
       }
 
+      // MU_CHANGE [BEGIN] - Allocate IO Queue Buffer
+      QueuePageCount = EFI_SIZE_TO_PAGES (Private->SqData[0].NumberOfEntries * 2^Private->SqData[0].EntrySize)
```
Reminder to fix these xors.

VivianNK commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.h:
```diff
@@ -66,4 +86,25 @@ NvmeIdentifyNamespace (
   );
 
+// MU_CHANGE [BEGIN] - Allocate IO Queue Buffer
+// TODO should this be used on reset or do we want to trust our private data structure?
```
@apop5 thoughts?

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -271,6 +276,22 @@ EnumerateNvmeDevNamespace (
 
+    if (Private->NumberOfDataQueuePairs > 1) {
+      // We have multiple data queues, so we can support the BlockIo2 protocol
+      Status = gBS->InstallProtocolInterface (
```
The UEFI spec recommends using `InstallMultipleProtocolInterfaces()` always, even for a single protocol: https://uefi.org/specs/UEFI/2.10/07_Services_Boot_Services.html?highlight=installprotocolinterface#efi-boot-services-installprotocolinterface

```
InstallMultipleProtocolInterfaces() performs more error checking than InstallProtocolInterface(), so it is recommended that InstallMultipleProtocolInterfaces() be used in place of InstallProtocolInterface()`
```

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -477,6 +509,29 @@ UnregisterNvmeNamespace (
+  //
+  // If BlockIo2 is installed, uninstall it.
+  //
+  if (Device->Controller->NumberOfDataQueuePairs > 1) {
```
Because BlockIo2 is not widely used, I would recommend trying to uninstall the other protocols first, since that's where real usage would be, then do BlockIo2 if required and not duplicate the OpenProtocol logic, just maintain if either uninstall failed

VivianNK replied:
Part of uninstallMultipleProtocols involves reinstalling all protocols if even one of them fails to uninstall.  to not have a split state. @apop5 can you confirm?
https://github.com/microsoft/mu_basecore/pull/1260#discussion_r1968459405

os-d replied:
That's not what I mean, what I mean is that you do the uninstall of BlockIo2 if required, if that fails open the protocol on the dummy interface, then go uninstall the more widely used protocols and then open the dummy interface if those fail. My suggestion is to not do the open protocol twice, but simply keep track of whether the BlockIo2 protocol uninstall failed and if either uninstall failed, then do the openprotocol. I.e. don't do that part twice.

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -288,12 +309,23 @@ EnumerateNvmeDevNamespace (
+        // MU_CHANGE [BEGIN] - Request Number of Queues from Controller
+        if (Private->NumberOfDataQueuePairs > 1) {
+          // We have multiple data queues, so we need to uninstall the BlockIo2 protocol
+          gBS->UninstallProtocolInterface (
```
Same with uninstall: https://uefi.org/specs/UEFI/2.10/07_Services_Boot_Services.html?highlight=installprotocolinterface#efi-boot-services-uninstallprotocolinterface

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -957,9 +1011,11 @@ NvmExpressDriverBindingStart (
   UINTN                               Bytes;
   EFI_NVM_EXPRESS_PASS_THRU_PROTOCOL  *Passthru;
-  // MU_CHANGE - Support alternative hardware queue sizes in NVME driver
-  UINTN  QueuePageCount = PcdGetBool (PcdSupportAlternativeQueueSize) ?
```
You are changing a previous MU_CHANGE, should this PR either be marked `[SQUASH ON REBASE]` or if completely superceding the previous commit(s), reverting that commit and applying this one as a `[REBASE & FF]`?

That might look like splitting this PR into multiple commits, one of which is squashed with the previous commit or reverting it.

VivianNK replied:
Reverted the original change then re-applied after my own

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -1033,9 +1089,41 @@ NvmExpressDriverBindingStart (
+    // Set the sizes of the admin submission & completion queues in number of entries
+    Aqa->Asqs  = PcdGetBool (PcdSupportAlternativeQueueSize) ? MIN (NVME_ALTERNATIVE_MAX_QUEUE_SIZE, Private->Cap.Mqes) : NVME_ASQ_SIZE;
+    Aqa->Rsvd1 = 0;
+    Aqa->Acqs  = PcdGetBool (PcdSupportAlternativeQueueSize) ? MIN (NVME_ALTERNATIVE_MAX_QUEUE_SIZE, Private->Cap.Mqes) : NVME_ACQ_SIZE;
```
I might be misremembering this change, but if we are now querying the NVMe controller for the queue size supported, should we not just use that, regardless of a PCD? Or is this for the case where a controller is wrong about its queue size?

In either case, it either seems to me like we don't need the PCD (and just take what the controller supports or the max we support) or that the HW needs to be fixed, but failing that we would need a PCD that tells us what queue size to use (but again I am dubious of not doing what the HW tells us).

VivianNK replied:
The PCD essentially changes what the maximum size we (the driver) support is. By default, the driver has hard-coded maximum sizes that vary depending on the queue type/which queue pair it belongs to. With the PCD there is one maximum size used for all queues.

This PR's change is regarding the number of queue pairs, but the PCD is concerning the driver's supported "size" of each queue in the number of queue entries per queue.

os-d replied:
Again, I know there is history here, but I don't remember it, why not just change what size the driver supports to the uniform size?

 I may be misunderstanding, but I thought part of the change is to query the HW for what queue size it supports? And then it doesn't matter what size the driver supports as long as we just do what the HW actually supports?

VivianNK replied:
For compatibility, we should keep the Pcd, and can contact the original requestors to see if it's still needed. For now, this change can be made independently.

os-d replied:
Sounds good to me, can you file an issue on mu_basecore to track removing the PCD (or at least finding out if it is still needed)?

VivianNK replied:
https://github.com/microsoft/mu_basecore/issues/1335

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -1033,9 +1089,41 @@ NvmExpressDriverBindingStart (
+    Private->CqData[0].EntrySize = NVME_IOCQES_MIN;
+
+    // Calculate the number of pages required for the admin queues
+    // TODO create helper functions for calculating num pages for a queue.
```
I would either implement the TODO or remove it. If this is the only place this calculation is done, I don't think the helper function is needed

VivianNK replied:
done

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -1033,9 +1089,41 @@ NvmExpressDriverBindingStart (
+
+    // Calculate the number of pages required for the admin queues
+    // TODO create helper functions for calculating num pages for a queue.
+    AdminQueuePageCount = EFI_SIZE_TO_PAGES (Private->SqData[0].NumberOfEntries * LShiftU64 (2, Private->SqData[0].EntrySize))
+                          + EFI_SIZE_TO_PAGES (Private->CqData[0].NumberOfEntries * LShiftU64 (2, CqData[0].EntrySize));
```
```suggestion
    AdminQueuePageCount = EFI_SIZE_TO_PAGES ((Private->SqData[0].NumberOfEntries * LShiftU64 (2, Private->SqData[0].EntrySize))
                          + (Private->CqData[0].NumberOfEntries * LShiftU64 (2, CqData[0].EntrySize));

```
This would potentially waste less memory since each calculation is rounding to the nearest page, but if the sizes together could fit within the same page, we would allocate an extra page.

VivianNK replied:
In `NvmeControllerInitAdminQueues` we have to write the base address of both the submission and completion queues.  Are you suggesting the base address of the completion queue starts in the middle of a page after the end of the submission queue? This deviates from previous behavior where each queue was allocated space in units of whole pages. In NvmExpress.c, this is described in a comment where for example, the 1st 4kB boundary is the start of the admin submission queue, and the 5th 4kB boundary is the start of the completion queue. 

Btw maybe the variable would be better named as "AdminQueuePairPageCount"

os-d replied:
Gotcha, I missed both queues are pulling from the same buffer.

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -1318,11 +1428,34 @@ NvmExpressDriverBindingStop (
       }
 
+      // MU_CHANGE [BEGIN] - Allocate IO Queue Buffer
+      QueuePageCount = EFI_SIZE_TO_PAGES (Private->SqData[0].NumberOfEntries * LShiftU64 (2, Private->SqData[0].EntrySize))
```
Same comment on allocation here

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.h:
```diff
@@ -67,6 +68,11 @@ extern EFI_DRIVER_SUPPORTED_EFI_VERSION_PROTOCOL  gNvmExpressDriverSupportedEfiV
 #define NVME_CCQ_SIZE  1                                // Number of I/O completion queue entries, which is 0-based
 
+// MU_CHANGE [BEGIN] - Allocate IO Queue Buffer
+#define NVME_IOSQES_MIN  6                              // Minimum I/O submission queue entry size (Figure 280)
```
Copied from the spec perhaps? Figure 280 doesn't seem to make sense here

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressBlockIo.c:
```diff
@@ -992,7 +1006,7 @@ NvmeBlockIoReset (
   Private = Device->Controller;
 
-  Status = NvmeControllerInit (Private);
+  Status = NvmeControllerReset (Private); // MU_CHANGE [BEGIN] - Allocate IO Queue Buffer
```
I haven't looked at the functions, but reset also does an init, presumably?

VivianNK replied:
Updated the commit message to detail the differences between init and reset: https://github.com/microsoft/mu_basecore/pull/1260/commits/5e8879cecaa96d988139401e46d9b6da97fd60e0

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -566,8 +632,94 @@ NvmeIdentifyNamespace (
+  using the NumberOfDataQueuePairs field.
+
+  @param  Private          The pointer to the NVME_CONTROLLER_PRIVATE_DATA data structure.
+  @param  Ndqpr             The number of data queue pairs requested.
```
```suggestion
  @param  Ndqpr            The number of data queue pairs requested.

```

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -566,8 +632,94 @@ NvmeIdentifyNamespace (
+
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "Set Features Command for Number of Queues failed with Status %r\n", Status));
+    return Status;
```
Do we need to set `Private->NumberOfDataQueuePairs` to a default value or is that handled elsewhere?

VivianNK replied:
In the case of an error during the Set Features command, would you suggest populating `Private->NumberOfDataQueuePairs` with 0? If there's a failure communicating with the controller, then we return a device error saying the device can't be started.

os-d replied:
No, that was just my question, whether that field needed to be set (i.e. did execution continue without this function, just assuming a default number of queues or did it fail out entirely).

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -709,6 +875,9 @@ NvmeCreateIoSubmissionQueue (
                                  );
     if (EFI_ERROR (Status)) {
+      // MU_CHANGE [BEGIN] - Request Number of Queues from Controller
+      DEBUG ((DEBUG_ERROR, "%a(%d): Create Submission Queue Command %d failed with Status %r\n", __FILE__, __LINE__, Index, Status));
```
In general `__FILE__` and `__LINE__` are only used for personal debugging and typically just `__func__` is used for checking in.

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -718,29 +887,170 @@ NvmeCreateIoSubmissionQueue (
+  EFI_STATUS  Status;
+
+  // Offset completion queue with submission queue size
+  AsqsPages = EFI_SIZE_TO_PAGES (Private->SqData[1].NumberOfEntries * 2^Private->SqData[1].EntrySize);
```
I honestly had no idea you could use `^` in C :). Can we convert this and other usages to the far more common shift operation?

VivianNK replied:
You can't! It's a bitwise xor, thought i fixed them all.

os-d replied:
That's what I thought :). Thanks!

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -718,29 +887,170 @@ NvmeCreateIoSubmissionQueue (
+  // buffer so we start at the beginning of that buffer.
+  //
+  ZeroMem (Private->DataQueueBuffer, EFI_PAGES_TO_SIZE (QueuePairPageCount) * Private->NumberOfDataQueuePairs);
+  for (Index = 0; Index < Private->NumberOfDataQueuePairs; Index++) {
```
Can we just start from Index = 1 in the loop and go to `<=`?

VivianNK replied:
Then half of the Index usages below would need to use Index-1 (for multiplying with QueuePairPageCount, since we want to start with 0)

os-d replied:
Gotcha, missed the right hand side usage of Index.

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -718,29 +887,170 @@ NvmeCreateIoSubmissionQueue (
+  UINTN       QueuePairPageCount;
+  EFI_STATUS  Status;
+
+  // TODO do we want to store the AsqsPages/QueuePairPageCount in the private data structure? so we don't have to recaclulate it later on?
```
Can the TODO either be addressed or removed?

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -718,29 +887,170 @@ NvmeCreateIoSubmissionQueue (
+  //
+  // Address of admin completion queue.
+  //
+  // MU_CHANGE [BEGIN] - Support alternative hardware queue sizes in NVME driver
```
Seems the whole function is a MU_CHANGE, can we move the tags to indicate that?

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -718,29 +887,170 @@ NvmeCreateIoSubmissionQueue (
+  //
+  // Address of admin submission queue.
+  //
+  Asq = (UINT64)(UINTN)(Private->BufferPciAddr) & ~0xFFF;
```
```suggestion
  Asq = (UINT64)(UINTN)(Private->BufferPciAddr) & ~EFI_PAGE_MASK;

```

And other places

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -718,29 +887,170 @@ NvmeCreateIoSubmissionQueue (
+  //
+  // Program admin submission queue address.
+  //
+  // TODO how can we program the data queue base address? Or do we need to?
```
Can we address or remove?

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -839,103 +1145,213 @@ NvmeControllerInit (
+  CopyMem (Sn, Private->ControllerData->Sn, sizeof (Private->ControllerData->Sn));
+  Sn[20] = 0;
+  CopyMem (Mn, Private->ControllerData->Mn, sizeof (Private->ControllerData->Mn));
+  Mn[40] = 0;
```
Some comments on what the setting of random values is?

VivianNK replied:
Added a comment in the Reset function (the new one), but this was copied from NvmeControllerInit. It's null-terminating the local serial number and model number variables for when they're printed a few lines later (https://github.com/microsoft/mu_basecore/commit/da7c72740f7f7172ffa0e62a512ad291ea0f4f87)

os-d replied:
Sure, I trust it's needed, but in 3 years when we come back to this we won't know what it is doing without comments :)

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -839,103 +1145,213 @@ NvmeControllerInit (
+  }
+
+  // Using the first data queue size for the number of pages required for the data queues
+  QueuePairPageCount = EFI_SIZE_TO_PAGES (Private->SqData[1].NumberOfEntries * 2^Private->SqData[1].EntrySize)
```
Same comments

VivianNK replied:
Re: allocating an extra page?

os-d replied:
Yes, but I thought I resolved this

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -839,103 +1145,213 @@ NvmeControllerInit (
     return Status;
   }
 
+  DEBUG ((DEBUG_INFO, "Private->DataQueueBuffer = [%016X]\n", (UINT64)(UINTN)Private->DataQueueBuffer));
```
Probably don't need this print checked in

VivianNK replied:
Copied from https://github.com/microsoft/mu_basecore/blob/dev/202405/MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c#L901
thinking it might be helpful info. Will defer to your judgement though.

os-d replied:
I don't think that provides any value in the log, it will just print the address of the buffer, which from a log doesn't tell you anything (other prints, queue size, whether BlockIo2 was installed or not, etc. do provide info from the log point of view).

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -839,103 +1145,213 @@ NvmeControllerInit (
+  UINT8                Sn[21];
+  UINT8                Mn[41];
+
+  DEBUG ((DEBUG_INFO, "%a: Begin Controller Reset\n", __FUNCTION__));
```
```suggestion
  DEBUG ((DEBUG_INFO, "%a: Begin Controller Reset\n", __func__));

```

os-d reviewed (commented):
This is a large PR that includes several different pieces, in order to prepare for upstreaming and making review easier, I would recommend splitting it up into at least a few different commits, including splitting commits per pkg (new NVMe industry standard definitions should be their own commit, refactoring of exsiting MU_CHANGE may be its own commit, get/set features another commit).

VivianNK committed:
MdeModulePkg: NvmExpressDxe: Allocate Io Queue buffer after setting
the number of queues with the controller.

This commit refactors NvmeControllerInit() in NvmExpressHci.c.
First it was modified to contain the logic for allocating the I/O queue
buffer after the number of queues has been set with the controller.
This change ensures that the I/O queue buffer is allocated with the
correct number of pages. It also now contains the logic for allocating
the Admin Queue buffer after getting the controller capabilites and setting
the Admin Queue Attributes.

Because NvmeControllerInit() was used for a controller reset,
a new function, NvmeControllerReset(), was created to only
apply the reset logic. For example, on reset the Admin Queue
Attributes should stay the same. The queue buffer allocations
can also remain the same. In the new reset function, more
checks are added to ensure the AQA and Number of Queues haven't changed.

Signed-off-by: Vivian Nowka-Keane <vnowkakeane@linux.microsoft.com>

VivianNK committed:
MdeModulePkg: NvmExpressDxe: Add get admin queue attributes function

VivianNK committed:
Revert "MdeModulePkg: Add alternative queue sizes in nvme driver to support specific devices"

This reverts commit 028442f3ba2d3d0e76df29d5046c1f7c51981418.

VivianNK committed:
MdeModulePkg: NvmExpressDxe: Use Mqes value to limit queue size

Signed-off-by: Vivian Nowka-Keane <vnowkakeane@linux.microsoft.com>

apop5 commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -1031,11 +1082,42 @@ NvmExpressDriverBindingStart (
+    //
+    // Set the Admin Queue Atttributes
+    //
+    Aqa = AllocateZeroPool (sizeof (NVME_AQA));
```
Memory Leak on Aqa?

 Its allocated and used, but it never appears to be freed, and it seems to go out of scope at the end of the call. 
 
 Maybe it should be a local variable?

VivianNK replied:
Good call, thanks! Made it a local variable.

os-d reviewed (dismissed):
I assume this will go in as as rebase & FF?

I still have questions on the alternative max queue size and different ways to handle that, but since this PR just reapplies the commit, we can defer that.

kuqin12 commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -288,12 +308,23 @@ EnumerateNvmeDevNamespace (
                );
+
+        // MU_CHANGE [BEGIN] - Request Number of Queues from Controller
+        if (Private->NumberOfDataQueuePairs > 1) {
```
This check is sprinkled in the module a few times. Would it be better to make it a macro so that we can change it more easily?

VivianNK commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.h:
```diff
@@ -78,7 +84,17 @@ extern EFI_DRIVER_SUPPORTED_EFI_VERSION_PROTOCOL  gNvmExpressDriverSupportedEfiV
+#define NVME_MAX_QUEUES  3
+
+// MU_CHANGE [BEGIN] - Request Number of Queues from Controller
+#define NVME_SUPPORT_BLOCKIO2(ContollerPointer)  (((ContollerPointer)->NumberOfDataQueuePairs) > 1)
```
@kuqin12 Added the macro here.

VivianNK commented:
Latest changes saved in this commit: https://github.com/VivianNK/mu_basecore/commit/bf7108f8d7ccf30f755b713015c654080a1e9522

Storing here since it will be squashed.

apop5 commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -818,19 +1119,82 @@ NvmeControllerInit (
+                    &Private->Mapping
+                    );
+  if (EFI_ERROR (Status) || (Bytes != EFI_PAGES_TO_SIZE (AdminQueuePageCount))) {
+    return Status;
```
When returning here, do the Private->Buffer need to be freed? Or does the calling code know that it is responsible for freeing it?

VivianNK replied:
The calling code is responsible for freeing the buffer. In NvmExpress.c:NvmExpressDriverBindingStart, if this returns an error, it will goto exit and free. BUT if there isn't an error and Bytes != AdminQueuePages, it wont' return an error, will continue, but should reach the Exit label and free the buffer.
So I think we should return an error here regardless, to go straight to Exit on return from this function.  I changed it to return `EFI_DEVICE_ERROR`. What do you think?

apop5 replied:
You are following what was done in NvmExpressDriverBindingStart, which seems fine.

apop5 commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -839,103 +1203,225 @@ NvmeControllerInit (
+                    );
+
+  if (EFI_ERROR (Status) || (Bytes != EFI_PAGES_TO_SIZE (QueuePairPageCount * Private->NumberOfDataQueuePairs))) {
+    return Status;
```
Same here, if the nvme controller isn't able to map the buffer that was allocated, will the caller code know about freeing the DataQueueBuffer?

VivianNK replied:
Same as other response ^

apop5 replied:
same as above response.

apop5 commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -839,103 +1203,225 @@ NvmeControllerInit (
+    return EFI_DEVICE_ERROR;
+  }
+
+  if ((VidDid[0] == EFI_PAGE_MASK) || (VidDid[1] == EFI_PAGE_MASK)) {
```
I don't know what this is trying to do with vendor id...why is it comparing these to EFI_PAGE_MASK?

VivianNK replied:
This comes from the commit here: https://github.com/microsoft/mu_basecore/commit/1e93d34be2ca33aa868460e0b3cda45a43cdabae#diff-e8680836e3bdb35921692e1fb5621eb8bec926e59953e6b9ad249cbfa1e63205R763

If the VidDid is 0xFFFFF it's not valid or wasn't read properly
@os-d suggested using EFI_PAGE_MASK instead of the hard-coded value.

os-d replied:
That is not what I meant :). This was `0xFFFF` not `0xFFF` for one thing, but also does not relate to the page mask. My comment was for the cases where you were checking something was page aligned or forcibly page aligning it.

![image](https://github.com/user-attachments/assets/dff884b1-afaa-4174-90e3-63f78ca66c85)

os-d replied:
Having this compared to a macro makes sense, but let's use the right one :). Which in this case would be creating an `INVALID_VID_DID` or whatever (or if something already exists). But because you are replaying the commit on top, I don't think it needs to change.

I did have a question about the replaying of that commit, I know that came from my comment asking if this should be squashed with that or if it was completely subsuming that, but the answer is no. So does it make it cleaner to do the revert and then the replay? If not, we can just put this one on top, I was just trying to understand, it was hard to tell from the original PR if this was completely changing that original commit or not.

VivianNK replied:
Added NVME_INVALID_VID_DID in NvmExpress.h

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -477,6 +509,24 @@ UnregisterNvmeNamespace (
+  //
+  BlockIo2Status = Status;
+  if (Device->Controller->NumberOfDataQueuePairs > 1) {
+    Status = gBS->UninstallProtocolInterface (
```
We want to use `UninstallMultipleProtocolInterfaces` here, too, right?

os-d replied:
And then here I would just directly use `BlockIo2Status` and not set it on line 524.

VivianNK replied:
Done

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -477,6 +509,24 @@ UnregisterNvmeNamespace (
+  //
+  // If BlockIo2 is installed, uninstall it.
+  //
+  BlockIo2Status = Status;
```
I would just set `BlockIo2Status` to `EFI_SUCCESS` above, so that if the preceding code changes there is no concern about what `Status` is at this point.

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.h:
```diff
@@ -78,7 +78,13 @@ extern EFI_DRIVER_SUPPORTED_EFI_VERSION_PROTOCOL  gNvmExpressDriverSupportedEfiV
+// Queue 1 - Blocking I/O (BlockIo Protocol)
+// Queue 2 - Asynchronous I/O (BlockIo2 Protocol)
+// MU_CHANGE [END] - Request Number of Queues from Controller
+#define NVME_MAX_QUEUES  3
```
nit: line 87 changed, too, so would move MU_CHANGE to after it

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressBlockIo.c:
```diff
@@ -169,6 +169,13 @@ NvmeRead (
   EFI_TPL                       OldTpl;
 
+  // MU_CHANGE [BEGIN] - Request Number of Queues from Controller
+  if (&Device->AsyncQueue == NULL) {
```
This can never be `NULL`, can it? AsyncQueue is a LIST_ENTRY, so the address of it would be in the Device struct? Should we be doing `IsListEmpty` instead?

os-d replied:
In the existing code they raise the TPL before checking if the AsyncQueue is empty, should we not check that and instead check about the number of queues? Or just take the TPL raise/lower hit and check the list there?

VivianNK replied:
This was an artifact of a bug that was introduced and then fixed in this PR. It led to dereferencing a NULL pointer. That's been fixed by always initializing the async list. Good catch, I removed these now

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressBlockIo.c:
```diff
@@ -255,6 +262,13 @@ NvmeWrite (
   EFI_TPL                       OldTpl;
 
+  // MU_CHANGE [BEGIN] - Request Number of Queues from Controller
+  if (&Device->AsyncQueue == NULL) {
```
Same comment here.

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressBlockIo.c:
```diff
@@ -1540,6 +1554,13 @@ NvmeBlockIoFlushBlocksEx (
   Device = NVME_DEVICE_PRIVATE_DATA_FROM_BLOCK_IO2 (This);
 
+  // MU_CHANGE [BEGIN] - Request Number of Queues from Controller
+  if (&Device->AsyncQueue == NULL) {
```
Same comment.

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -564,8 +566,94 @@ NvmeIdentifyNamespace (
+  // These numbers do not include the admin queues.
+  //
+  // This driver at maximum supports 2 pairs of data queues. So we will take the minimum of the requested and allocated values.
+  // UNDEFINED: what if the controller allocates more queues than requested? And we only allocate/support the number requested?
```
Do we have an answer to this? At a minimum, we should check the returned queues allocated against what we support (2 or 3) and assert if it is something else, then fallback to the max supported. I assume that if the device has more queues allocated than we support, that is okay, because we wouldn't used them.

VivianNK replied:
The answer I received: It's legal for a controller to allocate more queues and it's fine for the driver to ignore them.

I modified this portion of the code to check if we were allocated more than requested and to print a debug_info statement and use the number requested as the driver's number of allocated queues.

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -918,6 +1019,21 @@ NvmeControllerInit (
+  // The number of queues allocated is returned and stored in the controller private data structure
+  // using the NumberOfDataQueuePairs field.
+  //
+  Status = NvmeSetFeaturesNumberOfQueues (Private, NVME_MAX_QUEUES - 1);
```
Aren't you double subtracting one? You have MAX_QUEUES is 3, -1 here, which gets 2 passed into the func, but then in the func you have:

```
// Populate the Set Features Cdw10 and Cdw11 according to Nvm Express 1.3d Spec
  // Note we subtract 1 from the requested number of queues to get the 0-based value
  SetFeatures.Bits.Fid             = NVME_FEATURE_NUMBER_OF_QUEUES;
  NumberOfQueuesRequested.Bits.Ncq = Ndqpr -1;
  NumberOfQueuesRequested.Bits.Nsq = Ndqpr -1;
  ```
  
  So the number of queues requested ends up as 1? Instead of 2?

VivianNK replied:
NVME_MAX_QUEUES includes the admin queues, so -1 to get to 2 data queue pairs. 
The SetFeatures command uses 0-based values so 0 = 1 data queue pair, 1=2 data queue pairs, etc.
I just added a check for the parameter being 0 to return invalid parameter.

os-d replied:
Gotcha, I was missing that this didn't include the admin queue in one spot but does in the other.

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -173,7 +173,7 @@ EnumerateNvmeDevNamespace (
     Device->BlockIo.FlushBlocks = NvmeBlockIoFlushBlocks;
```
The commit message for MdeModulePkg: Allocate Io Queue buffer after setting is missing a word in the third sentence.

VivianNK replied:
Fixed

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.h:
```diff
@@ -159,9 +181,17 @@ struct _NVME_CONTROLLER_PRIVATE_DATA {
   // NumberOfDataQueuePairs = Nsqa = Ncqa
   //
-  UINT32    NumberOfDataQueuePairs;
+  UINT32                  NumberOfDataQueuePairs;
```
This didn't need to change?

VivianNK replied:
After the changes on lines 187-194, uncrustify formatting wants to line up NumberOfDataQueuePairs with SqData, etc.

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -173,7 +173,7 @@ EnumerateNvmeDevNamespace (
 
     // MU_CHANGE [BEGIN] - Request Number of Queues from Controller
-    if (Private->NumberOfDataQueuePairs > 1) {
+    if (NVME_SUPPORT_BLOCKIO2 (Private)) {
```
It seems like this macro change should just be in the first commit that added this logic? e.g. no point in two new MU_CHANGE commits that just change the same thing?

VivianNK replied:
fixed

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -276,15 +276,26 @@ EnumerateNvmeDevNamespace (
+               &gEfiDiskInfoProtocolGuid,
+               &Device->DiskInfo,
+               NULL
+               );
```
Similarly, these changes seem like they are fix ups, should they be in the original commit that changed this?

os-d replied:
But also, if we failed to install BlockIo2, should we really uninstall all the other protocols here? If we don't believe BlockIo2 is widely used, wouldn't it be better to simply not use it rather than not use everything?

VivianNK replied:
Removed the uninstall

VivianNK replied:
Merged the error message change into the original commit

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -951,6 +962,66 @@ NvmExpressDriverBindingSupported (
+  EFI_STATUS  Status;
+  UINTN       QueuePageCount;
+
+  if ((Private != NULL) && (Private->Mapping != NULL)) {
```
Rather than checking `Private` multiple times, you can check `Private` once at the top of the function and simply return if it is `NULL`.

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -951,6 +962,66 @@ NvmExpressDriverBindingSupported (
+    Status = Private->PciIo->FreeBuffer (Private->PciIo, QueuePageCount, Private->Buffer);
+
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_ERROR, "%a: FreeBuffer Buffer failed %r\n", __func__, Status));
```
Probably worth throwing an `ASSERT_EFI_ERROR` in these failure cases, these messages would be easy to miss in a log and it would be good to catch in a debug build that this is failing.

VivianNK replied:
done

VivianNK replied:
Added

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -951,6 +962,66 @@ NvmExpressDriverBindingSupported (
+
+  if ((Private != NULL) && (Private->Buffer != NULL)) {
+    QueuePageCount = EFI_SIZE_TO_PAGES (Private->SqData[0].NumberOfEntries * (UINTN)LShiftU64 (2, Private->SqData[0].EntrySize))
+                     + EFI_SIZE_TO_PAGES (Private->CqData[0].NumberOfEntries * (UINTN)LShiftU64 (2, Private->CqData[0].EntrySize));
```
Making this calculation a macro probably makes sense now, too, because you are using it in multiple places and across functions and expect the calculation to be the same, so we should ensure that. i.e. the macro should be used in every place you do the calculation.

os-d replied:
Also, this one is for the admin queues, right? If so, can we add a comment to clarify?

VivianNK replied:
Added the macros `NVME_SQ_SIZE_IN_PAGES` and `NVME_CQ_SIZE_IN_PAGES` and comments

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -994,13 +1065,12 @@ NvmExpressDriverBindingStart (
+  NVME_CONTROLLER_PRIVATE_DATA  *Private;
+  EFI_DEVICE_PATH_PROTOCOL      *ParentDevicePath;
+  UINT32                        NamespaceId;
+  // MU_CHANGE - Allocate IO Queue Buffer
```
When we remove edk2 code lines, we just comment them out to make the change clearer in future integrations

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -1073,45 +1143,6 @@ NvmExpressDriverBindingStart (
-    }
-
-    Private->BufferPciAddr = (UINT8 *)(UINTN)MappedAddr;
-
```
Same for all of this

os-d replied:
And please check everywhere, because it occurs multiple places

VivianNK replied:
Added back, commented out

VivianNK replied:
Done, but will re-request review

VivianNK replied:
Done, but re-requesting review

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -1219,13 +1250,13 @@ NvmExpressDriverBindingStart (
-  if ((Private != NULL) && (Private->Mapping != NULL)) {
-    PciIo->Unmap (PciIo, Private->Mapping);
+  // MU_CHANGE [BEGIN] - Allocate IO Queue Buffer
+  Status = NvmExpressDriverCleanUpQueues (Private);
```
I don't know the flow here, why does this happen on both start and stop? I assume that the flow for a reset would be to call stop (and so queues are cleaned) then start (so queues don't need to be cleaned again?). In fact, this seems dangerous because `NvmExpressDriverCleanUpQueues` doesn't set any of the stuff to `NULL` after freeing it, so you would end up double freeing? Are there code paths here to independently do start and stop?

VivianNK replied:
Stop wouldn't be called in this case because `Exit:` is only used in a failure during start.

VivianNK replied:
Changed `NvmExpressDriverCleanUpQueues ` to set things to `NULL` after freeing

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -951,6 +962,66 @@ NvmExpressDriverBindingSupported (
+  UINTN       QueuePageCount;
+
+  if ((Private != NULL) && (Private->Mapping != NULL)) {
+    Private->PciIo->Unmap (Private->PciIo, Private->Mapping);
```
Per the below comment, do we need to set all of these to NULL so they don't get used again?

VivianNK replied:
Done

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.h:
```diff
@@ -18,6 +18,22 @@
+  @param[in] Private                 The pointer to the NVME_CONTROLLER_PRIVATE_DATA data structure.
+
+  @retval EFI_SUCCESS                The NVM Express Controller is initialized successfully.
+  @retval Others                     A device error occurred while initializing the controller.
```
`inializ[ed|ing]` -> `reset[ing]`

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -686,11 +704,10 @@ NvmeSetFeaturesNumberOfQueues (
   // UNDEFINED: what if the controller allocates more queues than requested? And we only allocate/support the number requested?
   NumberOfQueuesAllocated.Uint32  = CommandPacket.NvmeCompletion->DW0;
-  Private->NumberOfDataQueuePairs = NumberOfQueuesAllocated.Bits.Nsq + 1;
+  Private->NumberOfDataQueuePairs = MIN (NumberOfQueuesAllocated.Bits.Nsq + 1, Ndqpr);
```
This also seems like a fix up to an earlier commit? Can you go through and ensure that later commits don't have fixups to earlier commits? I.e. that each commit is as expected going into the tree

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -855,6 +872,139 @@ NvmeCreateIoSubmissionQueue (
+
+  //
+  // Address of Data I/O submission & completion queue(s).
+  // We are using the same table of buffer pointers that the admin queus are in, so we start the table from Index + 1, but we have a separate
```
queues

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -855,6 +872,139 @@ NvmeCreateIoSubmissionQueue (
+  //
+  // Address of admin submission queue.
+  //
+  Asq = (UINT64)(UINTN)(Private->BufferPciAddr) & ~EFI_PAGE_MASK;
```
Should page alignment be checked in the allocation function instead?

VivianNK replied:
Talked offline, PciIo->Map() uses AllocatePages so using this mask isn't necessary. Removed

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -855,6 +872,139 @@ NvmeCreateIoSubmissionQueue (
+  //
+  // Address of admin completion queue.
+  //
+  Acq = (UINT64)(UINTN)(Private->BufferPciAddr + AsqsPages * EFI_PAGE_SIZE) & ~EFI_PAGE_MASK;
```
All Pages * EFI_PAGE_SIZE can use EFI_PAGES_TO_SIZE instead

os-d replied:
Although actually you only seem to be using AsqsPages as a size, so just do AsqsSize instead?

VivianNK replied:
What if the size is not aligned to a page? I thought this ensured we allocate enough pages, not just the correct number of bytes

os-d replied:
My point is that you are only ever using `AsqsPages * EFI_PAGE_SIZE`, so instead of sprinkling that calculation everywhere, you may as well call the variable `AsqsBufferLen` or something and have

```
AsqsBufferLen = EFI_PAGES_TO_SIZE (EFI_SIZE_TO_PAGES (Private->SqData[0].NumberOfEntries * (UINTN)LShiftU64 (2, Private->SqData[0].EntrySize)));
```

In this way you can align up to the next page size as needed, but still use it as a size everywhere, since that is all you are using it for.

VivianNK replied:
Updated but I'm still using AsqsPages

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -943,78 +1160,225 @@ NvmeControllerInit (
-  DEBUG ((DEBUG_INFO, "Sync  I/O Completion Queue (CqBuffer[1]) = [%016X]\n", Private->CqBuffer[1]));
-  DEBUG ((DEBUG_INFO, "Async I/O Submission Queue (SqBuffer[2]) = [%016X]\n", Private->SqBuffer[2]));
-  DEBUG ((DEBUG_INFO, "Async I/O Completion Queue (CqBuffer[2]) = [%016X]\n", Private->CqBuffer[2]));
+  // Serial Number and Model Number ire not null-terminated strings, but will be printed as ones.
```
`ire` -> `are`

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -943,78 +1160,225 @@ NvmeControllerInit (
+    return Status;
+  }
+
+  if ((Private->SqData[0].NumberOfEntries != Aqa.Asqs) || (Private->CqData[0].NumberOfEntries != Aqa.Acqs)) {
```
Should we re-init if there is a mismatch? Or does the spec guarantee that across reset this will be the same?

VivianNK replied:
The spec says
> The Admin Queue properties (AQA, ASQ, or ACQ) are not reset as part of a Controller Reset

This is just validation/error checking I added. Do you think it's not needed?

os-d replied:
I think the validation is good, that’s what I wanted to confirm, if the spec guaranteed they will not change. However, my question still stands, what if they do change? (Misbehaving device). Does returning a failure here cause a re-init to happen? Previously we would re-init everytime a reset happened, right? Seems like we may want to have a failure path here that re-inits if we fail to reset?

Because this is spec defined behavior we are relying on, I am okay if the answer is add asserts and just fail out gracefully, deal with it if we encounter a misbehaving device, but I want to make sure that is a conscious decision and that it is documented (comment or commit message) because it is a behavior change itself (though I’m not sure how big a behavior change compared to the old init function).

VivianNK replied:
Added ASSERT and a comment

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -1045,6 +1409,8 @@ NvmeControllerInit (
   // Dump NvmExpress Identify Controller Data
   //
   CopyMem (Sn, Private->ControllerData->Sn, sizeof (Private->ControllerData->Sn));
+  // Serial Number and Model Number ire not null-terminated strings, but will be printed as ones.
```
`ire` -> `are`

VivianNK committed:
MdeModulePkg: NvmExpressDxe: Remove page mask

Page mask was used but not needed since the allocation of the Nvme
queues buffer is done with PciIo->Map which uses AllocatePages() in
RootBridgeIoMap().

Signed-off-by: Vivian Nowka-Keane <vnowkakeane@linux.microsoft.com>

VivianNK commented:
@chris-oo I revert and re=apply your commit (1e93d34be2ca33aa868460e0b3cda45a43cdabae) in this PR. Please review those changes

apop5 commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c:
```diff
@@ -271,6 +277,23 @@ EnumerateNvmeDevNamespace (
+                      );
+      if (EFI_ERROR (Status)) {
+        DEBUG ((DEBUG_ERROR, "%a: Failed to install BlockIo2 protocol. Error %r\n", __func__, Status));
+        goto Exit;
```
At the Point that Blockio2 attempts to install, if an error occurs, BlockIo et all are still installed on the handle, but jumping to Exit will leave the name spaces for all the nvme controllers un enumerated.

What is the best way to handle this scenario? Do the BlockIo et all need uninstalled? Should the system downgrade the Number of Io pairs and allow it to limp along just as on BlockIo?

os-d replied:
Since BlockIo2 is not heavily used, probably trying to continue on (with an assert) makes the most sense?

VivianNK replied:
To Aaron's point, what would happen if the system still thought it supported BlockIo2 because there are more than 1 data queue pairs? With blockIo2 not installed would all of those requests fail rather than attempt the synchronous queue? Will dig to see if I can find the answer

VivianNK replied:
In NvmExpress.c ProcessAsyncTaskList, if any of the subtasks fail it won't process subsequent ones. I'm not sure if this is the desired behavior (to continue knowing async tasks will fail). Before this change, BlockIo2 protocol was installed with BlockIo and if failed, all protocols would be uninstalled. 

I think it makes sense to maintain that logic. If the host is able to allocate 2 data queues and expecting to use BlockIo2, then if its installation fails, we fail out (uninstalling all protocols like we would have previously). Falling back to using blocking I/O only would imply the host wants to continue knowing the expected async tasks would fail, and I'm not sure if that's the case. Thoughts @apop5 @os-d ?

VivianNK replied:
Updated implementation to uninstall the other protocols to maintain the logic discussed above.
Issue created requesting the ability to continue after failed install of BlockIo2: https://github.com/microsoft/mu_basecore/issues/1358

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.h:
```diff
@@ -78,7 +87,28 @@ extern EFI_DRIVER_SUPPORTED_EFI_VERSION_PROTOCOL  gNvmExpressDriverSupportedEfiV
+// Queue 2 - Asynchronous I/O (BlockIo2 Protocol)
+#define NVME_MAX_QUEUES  3
+
+#define NVME_SUPPORT_BLOCKIO2(ContollerPointer)  (((ContollerPointer)->NumberOfIoQueuePairs) > 1)
```
A comment here would be helpful (why is this test the test for blockio2 support, e.g.)

os-d commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -818,95 +1133,136 @@ NvmeControllerInit (
+
+  // MU_CHANGE [END] - Allocate IO Queue Buffer
+
+  Status = NvmeDisableController (Private);
```
A comment here would be helpful, too (i.e. why are we disabling the controller when init'ing? Presumably the spec says to do this). I know this was from the existing code, but since it is getting moved around, helpful to update.

Flickdm commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -566,8 +629,102 @@ NvmeIdentifyNamespace (
+  NVME_ADMIN_SET_FEATURES_NUM_QUEUES        NumberOfQueuesAllocated;
+
+  Status = EFI_SUCCESS;
+  if (Ndqpr == 0) {
```
Does Ndqpr have an upper limit?

VivianNK replied:
Spec-defined limit for the 0-based values of both Number of submission queues (NSQ) and number of completion queues (NCQ) is 65,534 (16 bits).
Source: Nvme Base Spec Section 5.27.1.5 Number of Queues

Flickdm commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -566,8 +629,102 @@ NvmeIdentifyNamespace (
+  NVME_ADMIN_SET_FEATURES_NUM_QUEUES        NumberOfQueuesRequested;
+  NVME_ADMIN_SET_FEATURES_NUM_QUEUES        NumberOfQueuesAllocated;
+
+  Status = EFI_SUCCESS;
```
NIT

```suggestion
  Status = EFI_SUCCESS;

```

VivianNK replied:
done

Flickdm commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -566,8 +629,102 @@ NvmeIdentifyNamespace (
+  // Note we subtract 1 from the requested number of queues to get the 0-based value
+  SetFeatures.Bits.Fid             = NVME_FEATURE_NUMBER_OF_QUEUES;
+  NumberOfQueuesRequested.Bits.Ncq = Ndqpr -1;
+  NumberOfQueuesRequested.Bits.Nsq = Ndqpr -1;
```
NIT

```suggestion
  NumberOfQueuesRequested.Bits.Ncq = Ndqpr - 1;
  NumberOfQueuesRequested.Bits.Nsq = Ndqpr - 1;

```

VivianNK replied:
done

Flickdm commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c:
```diff
@@ -732,15 +1039,23 @@ NvmeControllerInit (
-  UINT16               VidDid[2]; // MU_CHANGE - Improve NVMe controller init robustness
-  UINT8                Sn[21];
-  UINT8                Mn[41];
+  UINT16               VidDid[2];  // MU_CHANGE - Improve NVMe controller init robustness
+  // NVME_ASQ             Asq;
+  // NVME_ACQ             Acq;
+  UINT8                 Sn[21];
+  UINT8                 Mn[41];
```
Where do these magic numbers come from?

Flickdm replied:
Serial Number & Model number in this code are copied from the NVME device and are assumed to be 40 bytes in length with the 41st byte reserved for null terminator (index 40)

apop5 replied:
https://nvmexpress.org/specifications/

Nvme Express base Specification section  4.5 Identifier Format and Layout (Informative)
The strings provided by the identify response are non -null terminated.

VivianNK replied:
From Nvm Express Base Spec 4.3.2
> The Serial Number (SN, bytes 23:04) and Model Number (MN, bytes 63:24) are defined in the Identify 
Controller data structure. The values are ASCII strings assigned by the vendor.

VivianNK replied:
Line 1525 has a comment where we null-terminate the result

Flickdm commented on MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.h:
```diff
@@ -78,7 +87,30 @@ extern EFI_DRIVER_SUPPORTED_EFI_VERSION_PROTOCOL  gNvmExpressDriverSupportedEfiV
+// Returns the number of pages required for a submission/completion queue
+// The Indices are the same as above, 0 for admin, 1 for blocking I/O, 2 for async I/O.
+//
+#define NVME_SQ_SIZE_IN_PAGES(ControllerPointer, Index) \
```
Are any of these fields controlled by the device? I see based on this that we allocate data?
